modules/renter/hostdb/hostweight.go

```diff
@@ -41,104 +128,227 @@ func (hdb *HostDB) calculateHostWeight(entry modules.HostDBEntry) types.Currency
+
+	// Cap the total amount of downtime allowed based on the total number of
+	// scans that have happened.
+	allowedDowntime := 0.03 * float64(len(entry.ScanHistory))
+	if uptimeRatio < 1-allowedDowntime {
+		uptimeRatio = 1 - allowedDowntime
+	}
+
+	// Calculate the penalty for low uptime. Penalties increase extremely
+	// quickly as uptime falls away from 95%.
 	//
-	// The renter has control over how much collateral the host uses.
-	// Currently, this control is not implemented, so the hosts are penalized
-	// for setting very high collateral values. Once the renter is clamping the
-	// amount being spent on collateral, the hostdb can also clamp the amount
-	// of collateral being taken into account by the host, to optimize the
-	// host's score for the renter's needs.
-	if entry.Collateral.IsZero() {
-		// Instead of zeroing out the weight, just return the weight as though
-		// the collateral is 1 hasting. Competitively speaking, this is
-		// effectively zero.
-		return weight
-	}
-	weight = weight.Mul(entry.Collateral)
+	// 100% uptime = 1
+	// 98%  uptime = 1
+	// 95%  uptime = 0.91
+	// 90%  uptime = 0.51
+	// 85%  uptime = 0.16
+	// 80%  uptime = 0.03
+	// 75%  uptime = 0.005
+	// 70%  uptime = 0.001
+	// 50%  uptime = 0.000002
+	uptimePenalty := float64(1)
+	for i := float64(1); i > uptimeRatio && i > 0.80; i -= 0.01 {
+		uptimePenalty *= uptimeRatio
+	}
+	return uptimePenalty
+}
+
+// calculateHostWeight returns the weight of a host according to the settings of
+// the host database entry. Currently, only the price is considered.
+func (hdb *HostDB) calculateHostWeight(entry modules.HostDBEntry) types.Currency {
+	collateralReward := hdb.collateralAdjustments(entry)
+	pricePenalty := hdb.priceAdjustments(entry)
+	storageRemainingPenalty := storageRemainingAdjustments(entry)
+	versionPenalty := versionAdjustments(entry)
+	lifetimePenalty := hdb.lifetimeAdjustments(entry)
+	uptimePenalty := hdb.uptimeAdjustments(entry)
+
+	// Combine the adjustments.
+	fullPenalty := collateralReward * pricePenalty * storageRemainingPenalty * versionPenalty * lifetimePenalty * uptimePenalty
+
+	// Return a types.Currency.
+	weight := baseWeight.MulFloat(fullPenalty)
+	if weight.IsZero() {
+		// A weight of zero is problematic for for the host tree.
+		return types.NewCurrency64(1)
+	}
 	return weight
 }
+
+// ScoreBreakdown provdes a detailed set of scalars and bools indicating
+// elements of the host's overall score.
+func (hdb *HostDB) ScoreBreakdown(entry modules.HostDBEntry) modules.HostScoreBreakdown {
+	return modules.HostScoreBreakdown{
+		AgeAdjustment:              hdb.lifetimeAdjustments(entry),
+		BurnAdjustment:             1,
+		CollateralAdjustment:       hdb.collateralAdjustments(entry),
+		PriceAdjustment:            hdb.priceAdjustments(entry),
+		StorageRemainingAdjustment: storageRemainingAdjustments(entry),
+		UptimeAdjustment:           hdb.uptimeAdjustments(entry),
+		VersionAdjustment:          versionAdjustments(entry),
```

Comment:
do these functions need to be methods of the hostDB? If not, is there good reason to believe that they will eventually need to be?